Random split polytomy

hyanwong opened the pull request:
Fixes #809

hyanwong commented:
Currently only a draft, as there are basically no tests, rather a long method name (but I feel we need the word "random" in there, since AFAIK this is the only routine in tskit which involves randomness), and I only have a horribly inefficient way to check the time of mutations above child nodes. I'm sure the latter could be optimised - suggestions for approaches are welcome.

benjeffery commented:
Great stuff - haven't reviewed the code yet, but one question:is where we get the RNG on the C side (if indeed we intend to implement this C-side).

hyanwong commented:
What RNG is used in msprime on the C side? If we do want a C side implementation, I assume we just copy that. Having said which, profiling the code suggests (surprisingly) that the majority of the time is spent in `table.add_row()`, so a C implementation would mainly be to allow C users to access the functionality.

benjeffery commented:
> What RNG is used in msprime on the C side? 
GSL is used, which we can't use here due to licensing restrictions (It's GPL were MIT), hence it being an issue. One idea we had for another method was to pass in a random array - but I'm guessing here we don't know how many random numbers will be needed? In the other use case it was known.

grahamgower commented:
PCG family of random generators could be an option: https://www.pcg-random.org/download.html

hyanwong commented:
> PCG family of random generators could be an option: https://www.pcg-random.org/download.html

Numpy [currently uses](https://numpy.org/doc/stable/reference/random/index.html?highlight=random#quick-start) PCG64 as its default PRNG, which is good enough for me!

hyanwong commented on python/tskit/tables.py:
```diff
@@ -2781,3 +2782,204 @@ def union(
+                    nodes_changed.add(edge.parent)
+
+            if check_mutations and prev_tree is not None:
+                # This is grim. There must be a more efficient way.
```
@brianzhang01 - can you think of a nicer way to find the oldest mutation above a node?

hyanwong replied:
I just thought - we can probably do this by pre-sorting a copy of the mutations - sorting first by node, then by time (if there are no mutation times, then we don't need to do this anyway). The slight problem is that we might have mutations that are oldest above a node, but not on the current tree. We should be able to mask those out, however, by taking the list of mutations for this node and restricting to only the sites in this tree.

AdminBot-tskit commented:
📖 Docs for this PR can be previewed [here](https://tskit-dev.github.io/tskit-docs/815/)

hyanwong commented:
@benjeffery - I've revised this after discussion with Jerome, so that the code is placed in `combinatorics.py` and doesn't clog up `trees.py`. I've also removed the ability to modify the epsilon value, and now we just error out if epsilon is too large. There's a helpful error message if this is because of node times, and a less helpful one (at the end of the function) which traps messages about mutation times and raises an error mentioning epsilon. I'm not sure this is the right way to do it, but it avoids having to explicitly check for mutation times, which was a pain in the previous version. The RNG is the one from numpy, so porting this to C might be less obvious. The only "method" is "random" for the moment, but I think it's worth having that argument in as it emphasises the fact that by default there is a stochastic component to this algorithm.

I've added Jerome as a reviewer, as much of this was hist suggestion for reworking, but I don't know how much he wants to check over it all?

benjeffery commented:
I've just started reviewing.

jeromekelleher commented on python/tests/test_topology.py:
```diff
@@ -7951,3 +7953,163 @@ def test_star_branch_length(self):
+
+    # Statistical test coded here for reference
+    # @unittest.skip("Testing statistical properties is nontrivial in a unit test")
+    def test_equiprobable(self):
```
I don't think we can do this as a unit test as it'll be too slow and cause problems. We'll need some sort of validation suite or something. I'm not sure what to do with the code for now though.

@benjeffery, any thoughts?

hyanwong replied:
For the moment I can uncomment the skip and say "To be put in validation suite". How is this sort of simulation testing done in msprime?

hyanwong replied:
I didn't know about `@pytest.mark.slow` - seems like a good compromise.

jeromekelleher reviewed (commented):
It basically looks good, although I haven't really grokked the algorithm. This is quite a tricky thing, which I would imagine needs quite a lot of validation and testing to make sure it has the right properties in all situations.

My inclination would be to merge this ASAP but "undocument" it ahead of the upcoming release. That way, we can use it internally for our own projects without committing ourselves to any particular API or semantics, and also give us a chance to consider how we'll validate stuff like this which we can't just do simple unit tests on.

hyanwong commented on python/tskit/combinatorics.py:
```diff
@@ -720,6 +721,208 @@ def is_symmetrical(self):
+            if len(edges_from_node[parent]) >= 3:
+                child_edges = list(edges_from_node[parent])
+                # We have a previous polytomy to break
+                parent_time = tables.nodes.time[parent]
```
Ah - I wondered if accessing the attribute was costly (I had a variable like this before, but took it out as it seemed redundant). I'll put it back in, thanks.

hyanwong commented on python/tskit/combinatorics.py:
```diff
@@ -720,6 +721,208 @@ def is_symmetrical(self):
+        raise ValueError(f"method must be chosen from {allowed_methods}")
+    rng = np.random.default_rng(seed=random_seed)
+    tables = ts.dump_tables()
+    # Store existing table data before clearing
```
If I create `existing_edges = tables.edges.copy()`, is it still worth creating `existing_edges_right = existing_edges.right` for use in an inner loop? Or is accessing `existing_edges.right` just as performant?

jeromekelleher replied:
Currently ``existing_edges.right`` does create a full copy of the array data, so if it's used in an inner loop it's worth making the array all right.

benjeffery reviewed (commented):
I've not dug in deep to the algo - but here are a few comments.

hyanwong commented:
I've made the review changes suggested by @benjeffery and @jeromekelleher . Do you want me to undocument this too (how? I could put the documentation into the `combinatorics.split_polytomy()` function, which I don't think currently gets exported to the docs).

benjeffery commented:
What's been the process before? Do we wait for the C port to document? Seems it is still useful in current form - maybe with a warning about perf.

hyanwong commented:
> What's been the process before? Do we wait for the C port to document? Seems it is still useful in current form - maybe with a warning about perf.

I don't think it's the performance that Jerome is worried about. I think it's the complexity of the algorithm. He doesn't want to document it without testing more extreme cases, I would guess.

Having said which, I have been running this algorithm on a variety of large simulated and inferred tree sequences without issue so far, although I haven't been doing explicit statistical tests for equiprobability on the resolved trees.

jeromekelleher commented:
> What's been the process before? Do we wait for the C port to document? Seems it is still useful in current form - maybe with a warning about perf.

It's not so much about perf as semantics - I'd need to see quite a lot more testing of how this works on various tree sequence topologies to be assured that it's actually dong the right thing in all cases. Things like nodes sequentially jumping out of a large polytomy tree-by-tree - how would this be resolved? How *should* it be resolved? Maybe we're happy that marginally we're resolving individual trees such that we're sampling from the space of binary trees, but it's not obvious that what we're doing is the correct thing across *all* trees in the sequence. What does it really mean, anyway, to resolve a polytomy that spans multiple trees, but has different nodes in each. This is not a straightforward problem and it needs real research to solve it, and I don't want to give the impression that we've *done* this research and that people should trust the code to do the right thing in all cases (which, is the goal: people should trust tskit).

So, this discussion makes me wonder: what if instead of trying to randomly resolve a whole tree sequence, we just did it tree-by-tree instead? I.e., the Tree class had a method ``resolve_polytomies`` that returned a Tree with the same interval, with all the poloytomies broken. The underlying tree sequence would consist of the same set of nodes, and would only have edges spanning the interval in question (so, if the tree is in the middle of the sequence, it will have missing data on both sides of it).

This would get rid of the tricky semantic questions, and make this much easier to test and validate. I'm pretty sure this fulfils your use case @hyanwong, since you're doing pretty non-standard stuff with the tree metrics, anyway (although it means you have to coiterate tree-by-tree, but that's OK: we could promote [this function](https://github.com/tskit-dev/tskit/blob/main/python/tests/tsutil.py#L1287) to be part of the public API. The performance impact of doing KC on a set of trees vs tree sequences would be entirely swamped by the cost of resolving the trees, anyway.

benjeffery commented:
@jeromekelleher Just a note for future in case this happens to someone else. Yan was not getting coverage reports due to following the repo on his personal circle CI account: https://support.circleci.com/hc/en-us/articles/360008097173-Why-aren-t-pull-requests-triggering-jobs-on-my-organization-

hyanwong commented:
> What does it really mean, anyway, to resolve a polytomy that spans multiple trees, but has different nodes in each. 

That should never happen. As soon as any of the nodes underneath a polytomy change (or the nod itself changes), we do a brand-new random resolution.

jeromekelleher commented:
> That should never happen. As soon as any of the nodes underneath a polytomy change (or the node itself changes), we do a brand-new random resolution. The complexity in the algorithm is simply spotting whether the edges under a polytomy have changed or not, when we move to a new tree (as the change could be in a completely different part of the tree)

Then why bother trying to do this at the tree sequence level? It seems like this would a lot simpler if we just regarded it on a marginal tree basis.

hyanwong commented:
> Then why bother trying to do this at the tree sequence level? It seems like this would a lot simpler if we just regarded it on a marginal tree basis.

Because we have very large number of samples, and hence a change to a new tree will probably not involve any changes to a polytomous node. I expect many of these polytomies to span a large number of trees without changing their children at all (the tree change could be lower down in the subtree, or in a different subtree). If we go tree-by-tree we re-break these every time.

jeromekelleher commented on python/tskit/combinatorics.py:
```diff
@@ -720,6 +721,207 @@ def is_symmetrical(self):
+
+    for interval, e_out, e_in in ts.edge_diffs(include_terminal=True):
+        for edge in itertools.chain(e_out, e_in):
+            if edge.parent != tskit.NULL:
```
``edge.parent`` can't be NULL if these are unmodified edges from the tree sequences

Same below

hyanwong replied:
Oh, good point. I always think that the root has a edge with a null parent, but that's not how we do it.

jeromekelleher commented:
I'm nervous about his code because it's a complicated algorithm that I don't understand, that we've only tested on a handful of topologies, haven't extensively validated and proved no theoretical properties about. The fact that we've jumped over the simpler case of doing this tree-by-tree makes me more nervous still.

Surely it makes sense to start with the simpler case of single-tree random resolution and then *if* this doesn't do what's needed we extend to the tree sequence case? We have more pressing things to do than worry about the precise details of the splitting semantics in tree sequences  - but if you want to add functions to the library then that's precisely what we have to do.

hyanwong commented:
Well, I'm mainly trying to get this in so that I can test the inferred trees from tsinfer. Why not include it but undocumented it, as you previously suggested? It seems like quite a lot more coding work to me to implement a Tree.split_polytomy() method, and applying such a thing treewise will also mean that operations on the resulting ts might be quite a lot slower, as there could be loads more edges, and it's already slow calculating the KC on the trees.

If it's not in tskit, then I guess I'll just include the code as-is in the tsinfer / tsdate testing codebase, but the reason for including this that I had hoped to slim that down a bit.

hyanwong commented:
I've just been thinking about how this would be done on a tree-by-tree basis, @jeromekelleher, and I don't think the algorithm would be much simpler, would it? You would loop over the trees, and every time you came across a node with >2 children, you would "cut" the edges leading to that node into 3 parts, one before the tree, one over the tree span, and another after the current tree. So you would either have to co-iterate over the edge diffs and the trees, and build up a set of new edges as you go, or do somewhat like I'm doing now and iterate over only the edge diffs and keep track of the number of children for each node using the diffs. It's all going to involve keeping track of currently used edges, in a slightly convoluted way. I think the algorithm would end up being almost as intricate.

Alternatively, you could simply cut up *all* the trees into separate edges, split the polytomies, and squash them together again, but for large trees this would be very painful, wouldn't it?

I feel like I'm missing an obvious way to do this, though. Do you or @benjeffery have any better suggestions?

jeromekelleher commented:
I was suggesting adding a ``Tree.split_polytomies`` method that just returns one tree:

> I.e., the Tree class had a method resolve_polytomies that returned a Tree with the same interval, with all the poloytomies broken. The underlying tree sequence would consist of the same set of nodes, and would only have edges spanning the interval in question (so, if the tree is in the middle of the sequence, it will have missing data on both sides of it). 

So, there's only one "real" tree in the underlying tree sequence (or ts.num_trees is at most 3). This is simpler because it completely avoids having to worry about the global semantics of the splitting operations, you just think about the local tree. That's something that's pretty well established, we understand how to do and we can easily document.

For your application, I guess you'd do something like
```python
for interval, source_tree,  inferred_tree in tskit.coiterate(source_ts, inferred_ts):
      distances = np.array(num_random_resolutions)
      for j in range(num_random_resolutions):
          resolved_tree = inferred_tree.resolve_polytomies()
          distances[j] = source_tree.kc_distanc(resolved_tree)
     mean_distance = np.mean(distances)
     # Now weight by either the interval's span or the number of sites in the interval and add to the 
     # total
```

This is nice and simple, and easy to describe.

hyanwong commented:
Oh, I see. So I never get a bifurcating tree sequence out of it at all. You are right, that would be easier. But then I wouldn't be able to store the resulting ts at all, unless I output all the trees as newick, or something. It's also convenient to be able to re-iterate over the split tree sequence a few times, to run various other tests on the bifurcating trees (e.g. using the dendropy  or R libraries library to get additional distance metrics, etc.).

If it's turning out to be a pain getting this in, then I guess I'm just going to stick with packaging it separately, TBH. I feel it would be too much work figuring out how to construct the individual trees when I have something that I'm happy with at the moment.

jeromekelleher commented:
@hyanwong - see my proposal over in #1013. Having done some reading on this (there's a significant combinatorial literature on it), it's not obvious to me that your algorithm is correct either - what are you basing your approach on?

hyanwong commented:
Thanks @jeromekelleher . The algorithm is the one used in Dendropy (now corrected, see e.g. https://github.com/jeetsukumaran/DendroPy/commit/7131bf0d2b3623667934581897d3cdf42cdadb65), I've discussed it in more detail with Mark Holder, and also (briefly, as an aside) in a email discussion with Emmanual Paradis, the author of the Ape phylogenetic package which provides this functionality in R. I can paste the correspondence below if you like. I'm sure it's correct.

hyanwong commented:
p.s. I've just noticed that the algorithm I use is that labelled 2.1 in Furnas (1984): https://link.springer.com/article/10.1007/BF01890123

jeromekelleher commented:
Fair enough, but the random binary tree algorithm is still buried heavily in implementation detail. I think the approach in #1013 is a lot easier to follow and verify.

So, my preferred approach here would be to use the single tree approach defined in #1013, combined with the tests and UI refinements you have here. We can extend to a "tree sequence" resolution in a follow up if that turns out to be necessary. Is this OK? If you agree I'll add some commits onto this PR to do this, so that the tree sequence approach is still in the git history, should we need it.

hyanwong commented:
OK. I think it's worth testing the algorithm for equiprobability, though: as you say, translating these correctly isn't always easy. My version has the benefit that you can allocate all the random numbers you need in one go, which is handy for python, but not an issue for C. I think Furnas (1984) shows that it has nice scaling properties (which is presumably why it was picked for other phylogenetic programs), but that might be true of Remy's method too (do you have a citation for this in the phylogenetic literature, as not everyone has access to TAOCP?).

jeromekelleher commented:
The reference is [here](http://www.cs.utsa.edu/~wagner/knuth/fasc4a.pdf), page 16. It's a linear time algorithm (I don't see how one can do much better?) and has been proved to sample uniformly from the space of labelled binary trees.

jeromekelleher commented:
I've updated this @hyanwong, can you take a look please?

I think it's mostly ready to go, but I think we should probably change the semantics of ``epsilon`` to do the following. If it is not specified, we space the node times out evenly along the branches of the randomly generated binary tree, such that epsilon we space out the nodes more evenly. We can easily do this now, and it will do what people want most of the time, I'd imagine.

If you really want a bunch of super-short branch lengths then you can specify epsilon manually.

hyanwong commented:
> I've updated this @hyanwong, can you take a look please?

Will do. I haven't followed the logic exactly yet, through.

> I think it's mostly ready to go, but I think we should probably change the semantics of `epsilon` to do the following. If it is not specified, we space the node times out evenly along the branches of the randomly generated binary tree, such that epsilon we space out the nodes more evenly. We can easily do this now, and it will do what people want most of the time, I'd imagine.

I started out thinking about an adaptive epsilon like this, but it's hard to do because of mutation times, I think? There's no easy way to get the time of the oldest mutation above a node, without actually just going through the mutations one by one. I think we said that would probably not be too much of a hit, but it seemed easier to just require epsilon to be set manually.

jeromekelleher commented:
> I started out thinking about an adaptive epsilon like this, but it's hard to do because of mutation times, I think? There's no easy way to get the time of the oldest mutation above a node, without actually just going through the mutations one by one. I think we said that would probably not be too much of a hit, but it seemed easier to just require epsilon to be set manually.


It's easy enough to get the sites and mutations - we just iterate through them all for the tree. We're already doing a O(n) traversal of the nodes, so there's no point in worrying about efficiency there.

hyanwong commented on python/tskit/combinatorics.py:
```diff
@@ -28,12 +28,133 @@
+        implementation of Algorithm R, because we are interested in
+        the leaf node labellings.
+
+        The pre-fasicle text is available here, page 16:
```
I've never heard of a fasicle?! Looks like it has an extra c in the spelling anyway!

hyanwong commented:
> > I started out thinking about an adaptive epsilon like this, but it's hard to do because of mutation times, I think? There's no easy way to get the time of the oldest mutation above a node, without actually just going through the mutations one by one. I think we said that would probably not be too much of a hit, but it seemed easier to just require epsilon to be set manually.
> 
> It's easy enough to get the sites and mutations - we just iterate through them all for the tree. We're already doing a O(n) traversal of the nodes, so there's no point in worrying about efficiency there.

OK, so we'll presumably have to add that logic and space the nodes out evenly between the oldest mutation above a child, or, if there are no mutation times, the oldest child node time?

hyanwong commented on python/tskit/combinatorics.py:
```diff
@@ -28,12 +28,133 @@
+        ts = tables.tree_sequence()
+    except tskit.LibraryError as e:
+        msg = str(e)
+        if msg.startswith(
```
If we are checking on the time of children or mutations above children, then I think it would be better to raise these errors when we know we are going to violate the time constraints, as then we can say which nodes or mutations are causing the problems, which is much more helpful to the user.

I didn't do this for mutation times because I removed the code for checking mutation times, but if we are going to reinstate it, we should raise the error there.

brianzhang01 reviewed (commented):
I had these comments queued up from a review I started but didn't finish. Hope the comments still show up (I don't even know what they are at this moment, haven't figured out how to view them).

hyanwong commented on python/tskit/combinatorics.py:
```diff
@@ -28,12 +28,133 @@
+            if x.parent is not None:
+                index = x.parent.children.index(x)
+                x.parent.children[index] = internal
+            rng.shuffle(internal.children)
```
I must admit that I don't quire understand why this extra shuffle is in there? Won't the first `rng.choice` make the selection random anyway?

hyanwong replied:
(by the way, I think this is essentially the same edge-addition algorithm that I originally implemented)

jeromekelleher replied:
We choose a node x to modify randomly in the first step, adding a new internal node which will be the parent of x and the n'th leaf. We must randomise the order of [x, n] because the leaf would aways be either the left or right child, which would (I think) mean we don't generate all possible trees.

I'll add a few comments to clarify, as it's a simple algorithm and worth explaining for future reference.

hyanwong replied:
I think it *does* generate all possible trees, even if the child order isn't shuffled. At least, it still did in my tests. And from reading the Knuth explanation, it doesn't seem necessary (I think?)

brianzhang01 commented:
Having read through the discussion...
- I am partial towards Jerome's suggestion to resolve individual trees on-the-fly. It's simpler and the KC metric operates tree-wise and then aggregates, so it would make for conceptually simpler KC code as well. The maintenance burden would be lower as more people would understand the code. Examples of maintenance include non-random resolution and different distributions of randomly resolving trees (e.g. rtree vs. rcoal resolution in ape: https://www.rdocumentation.org/packages/ape/versions/5.4-1/topics/rtree, current behavior is rtree)
- It seems like Yan's function would still have a nice place to live: tsinfer. Besides metrics, what is nice about having and serializing a tree sequence with polytomies resolved?
- The coiterate function in tsutil looks useful for metrics purposes
- For what it's worth, I currently have a KC with random resolution of polytomies that more-or-less resolves trees-on-the-fly and computes the KC between two trees at a particular position. Then I sample many positions and compute a Monte Carlo-style average. It's written in my own C++ codebase. The point here is that to cut down on the time of computing a tree-sequence-wide KC, this might be an option, but you'd want to measure the effects of the Monte Carlo approximation.

hyanwong commented:
> Having read through the discussion...
> 
> * I am partial towards Jerome's suggestion to resolve individual trees on-the-fly. It's simpler and the KC metric operates tree-wise and then aggregates, so it would make for conceptually simpler KC code as well. 

But I don't that's what the KC metric on a tree sequence does, @brianzhang01. Daniel coded up an an incremental algorithm, didn't he? I'm afraid that I still think that it's a bit nuts to re-resolve the polytomy every time anything anywhere on the tree changes: it doesn't feel like a natural thing to do, because the tree boundaries are really nothing to do with the polytomy itself. I suspect that it also won't scale to large trees, where pretty much every base could cause a tree change *somewhere* in the TS, probably miles away from the polytomy you are trying to resolve?

brianzhang01 commented on python/tests/test_combinatorics.py:
```diff
@@ -976,3 +977,30 @@ def test_forward_time_wright_fisher_simplified(self):
+    """
+
+    @pytest.mark.parametrize("n", [2, 3, 5, 10])
+    def test_random_binary_tree(self, n):
```
Maybe add a comment that we are checking for a valid binary tree on n leaves, and not checking anything about the randomness

jeromekelleher replied:
Done, good call.

brianzhang01 commented on python/tskit/combinatorics.py:
```diff
@@ -28,12 +28,133 @@
+    label = attr.ib(default=None)
+
+    @staticmethod
+    def random_binary_tree(leaf_labels, rng):
```
Should we document the arguments, leaf_labels and rng?

jeromekelleher replied:
Done. I've not done the formal sphinx params as this isn't part of the public API.

brianzhang01 commented on python/tskit/combinatorics.py:
```diff
@@ -28,12 +28,133 @@
+        raise ValueError(f"Method must be chosen from {allowed_methods}")
+
+    tables = tree.tree_sequence.dump_tables()
+    tables.keep_intervals([tree.interval], simplify=False)
```
This seems not ideal for a long tree sequence, but I guess there's no easy way around it? Instead of O(E) where E is the total number of edges, seems like there should be an O(K + log(E)) algorithm where K is the number of edges to keep, via stabbing queries? I'm not that familiar with the tskit backend though.

jeromekelleher replied:
This is a great point @brianzhang01. Currently this operation is O(E) as you say, and it's one of the things we could make more efficient if we had implemented stabbing queries. See #684 and other [related issues](https://github.com/tskit-dev/tskit/projects/5).

For this application, I think the current approach is fine, but we would like to speed it up at some point.

brianzhang01 commented on python/tskit/trees.py:
```diff
@@ -2370,6 +2370,67 @@ def kc_distance(self, other, lambda_=0.0):
+        :math:`(2n - 3)!!` possible bifurcating topologies are produced with equal
+        probability.
+
+        .. fixme:: Looks like this function is wrong
```
?

jeromekelleher replied:
What's the correct number here? (2n - 3)!! is clearly very wrong indeed, and (2n - 3)! is just wrong.

hyanwong replied:
Good point, sorry. Don't know how that snuck in. https://www.jstor.org/stable/2412810?seq=2 : 
<img width="88" alt="Screenshot 2020-11-19 at 23 25 37" src="https://user-images.githubusercontent.com/4699014/99736587-8ea0fc80-2abe-11eb-8f96-5333757c4864.png">

jeromekelleher replied:
Ah, [here](https://oeis.org/A005373/a005373.pdf) it is. We're missing a bunch of terms..

hyanwong replied:
I beat you to it :)

brianzhang01 replied:
I think these two expressions are the same, actually, by the conventions of double factorial. https://en.wikipedia.org/wiki/Double_factorial

hyanwong replied:
Oh yes. I must have looked that up and forgotten I had done so! Thanks @brianzhang01 . The double factorial thing is nice and concise.

brianzhang01 commented on python/tskit/trees.py:
```diff
@@ -2370,6 +2370,67 @@ def kc_distance(self, other, lambda_=0.0):
+        .. fixme:: Looks like this function is wrong
+
+        The returned :class`.Tree` will have the same genomic span as this tree,
+        and node IDs will be converved (that is, node ``u`` in this tree will
```
converted

brianzhang01 replied:
The idea of node IDs being converted isn't obvious to me, could use another sentence of documentation. So if I had leaves 0, 1, 2, and a polytomy 3 that is the parent of all three leaves, then in the returned tree I would have 0, 1, 2, 3, 4? Where 3 is the "converted" node ID and 4 is an extra node from breaking the polytomy?

brianzhang01 commented on python/tskit/trees.py:
```diff
@@ -2370,6 +2370,67 @@ def kc_distance(self, other, lambda_=0.0):
+            parent of the edge on which they lie. If ``epsilon`` is not small enough,
+            compared to the distance between a polytomy and its oldest child (or oldest
+            child mutation) these requirements may not be met. In this case an error is
+            raised, recommending a smaller epsilon value be used.
```
An alternative to epsilon would be to use https://numpy.org/doc/stable/reference/generated/numpy.nextafter.html. That should give finer resolution than trying to add a small floating point value.

hyanwong replied:
I didn't know about nextafter. This is nice, but I'm cautious about using it, because we have been bitten in the past by creating nodes fractionally smaller than something, and then not being able to wedge in extra stuff later on (e.g. mutations). I can't see why you might want to do this in this case, but I'm still wary. Anyway, Jerome's suggestion of defaulting to an even spread of times seems good to me.

jeromekelleher replied:
I agree with Yan here - if folks want to have a super-fine epsilon then they can provide it as a parameter, but as a default nextafter is a bit tricky. In practise 1e-10 is a reasonable default, I think, leaving enough space in between nodes for mutations (if they are needed) while still being very small.

I've changed my mind on the auto-spacing after thinking more about it @hyanwong - it'll be quite hard to test, and actually you probably *do* want super short branches on the randomly resolved polytomies.

brianzhang01 commented:
> But I don't that's what the KC metric on a tree sequence does, @brianzhang01. Daniel coded up an an incremental algorithm, didn't he? I'm afraid that I still think that it's a bit nuts to re-resolve the polytomy every time anything anywhere on the tree changes: it doesn't feel like a natural thing to do, because the tree boundaries are really nothing to do with the polytomy itself. I suspect that it also won't scale to large trees, where pretty much every base could cause a tree change _somewhere_ in the TS, probably miles away from the polytomy you are trying to resolve?

I see, I just skimmed #548 a bit. I guess I meant that the clearest mathematical definition of tree-sequence KC is as an average over the tree. But OK, if you want to use Daniel's function with polytomy breaking then I guess you would need a resolved tree-sequence.

> Yes, I didn't know about this. It's great! Should it be extended to multiple tree sequences, do you think?

I can't think of a clear use case off the top of my head.

jeromekelleher commented:
Thanks for the comments @hyanwong and @brianzhang01, this is all very helpful. I'll have to leave this for a few days, so will respond to your individual comments when I can get backt o it.

However, I'm pretty solidly against the idea of doing resolutions tree-sequence-wide now as i'm *sure* there's a bunch of complexities we won't have considered, and we don't need it right now.

AdminBot-tskit commented:
📖 Docs for this PR can be previewed [here](https://tskit-dev.github.io/tskit-docs/PR-815/)

AdminBot-tskit commented:
📖 Docs for this PR can be previewed [here](https://tskit-dev.github.io/tskit-docs/PR-815/)

jeromekelleher commented:
This is ready to go, I think. Here's some evaluation code, for the record.

```python
import tskit
import io

import numpy as np
import seaborn as sns
import pandas as pd
import matplotlib.pyplot as plt


def num_leaf_labelled_binary_trees(n):
    """
    Returns the number of leaf labelled binary trees with n leaves.

    https://oeis.org/A005373/
    """
    return int(np.math.factorial(2 * n - 3) / (2 ** (n - 2) * np.math.factorial(n - 2)))


for n in range(2, 7):
    N = num_leaf_labelled_binary_trees(n)
    print(f"Running n = {n}, {N} topologies")
    ranks = []
    tree = tskit.Tree.generate_star(n)
    for seed in range(1000 * N):
        split_tree = tree.split_polytomies()
        ranks.append(split_tree.rank())
    df = pd.DataFrame({"rank": ranks})
    assert len(df["rank"].unique()) == N
    ax = sns.countplot(x="rank", data=df)
    plt.savefig(f"n={n}.png")
    plt.clf()
```

![n=3](https://user-images.githubusercontent.com/2664569/99882982-14868a00-2c1c-11eb-9042-dcf3fdc0fbea.png)

![n=4](https://user-images.githubusercontent.com/2664569/99882985-194b3e00-2c1c-11eb-93f2-158b751c6726.png)

![n=5](https://user-images.githubusercontent.com/2664569/99882989-20724c00-2c1c-11eb-8eee-bef93faf9115.png)

hyanwong commented:
Thanks @jeromekelleher - did you remove the extra shuffle? You can probably use your tests above to convince yourself that it isn't needed.

jeromekelleher commented:
> Thanks @jeromekelleher - did you remove the extra shuffle? You can probably use your tests above to convince yourself that it isn't needed.

No, I left it in despite not being able to see any difference in having it from the tests. According to Knuth, there are 4n - 2 ways of creating a binary tree with n labelled leaves from one with n - 1. But, we're only choosing from 2n - 1 when we select the focal node x, so it seems to me we have to have the shuffle to get the right number.

As I say, I couldn't see any difference in the tests, but maybe there's something subtle in the leaf labellings that we're missing - we're only ever testing on the ordered labels, so maybe something would show up if we spent more time on that. But, there didn't seem much point to me, since the shuffle can't do any harm and there's not actually any good reason from the algorithm description's perspective for leaving it out.

hyanwong commented:
Sorry, just picking up a comment I made above @jeromekelleher - if possible it's *really* helpful to give some feedback about (a) what epsilon value to use, and (b) where the problem is when epsilon is too large. For node times (but not mutation times) we should be able to do that during the split algorithm, rather than right at the end of the function, when we have lost the information about where the problem is.

jeromekelleher commented:
Can you open an issue to track this please @hyanwong?

hyanwong commented:
Just to follow up. Switching to splitting polytomies per tree rather than per polytomous node now means that splitting polytomies is now taking longer than the inference process itself (and with better tree sequence, I suspect it will end up taking up more and more of the time, so I am still pretty keen on considering a tree-sequence-wide polytomy resolution algorithm.

hyanwong commented:
> Can you open an issue to track this please @hyanwong?

Opened at #1088
